### PR TITLE
C++ code compilation for pip wheel

### LINKED ISSUE #75: Code compilation during pip install
The C++ code needs to be compiled during pip installation instead of on the fly.

### PR DESCRIPTION
**Pull request by NikoOinonen**
Fixes #75 

Adds `setup.py` and `MANIFEST.in` in order to include the compiled C++ extensions into the pip wheel. The C++ source files are modified to be compatible with the Microsoft C++ compiler, so that the pip wheel can also be built on Windows.

In order not to make the C++ extension recompile every time by default, added an environment variable `PPAFM_RECOMPILE` that needs to be set to a non-empty value if you want the code recompile.

Also some reorganization to move non-essential Python modules and C++ source files to a new submodule called `dev` inside `ppafm`. This module is excluded from the pip wheel.

### CODE REVIEW

**Comment by ondrejkrejci** [comment]
Before deep-diving into the code, I have tried to download it and install it:
`pip install -e .` went succesfully, but of cousre the compilation of C++ did not went through as I do not have MS Visual Studio.

I was thinking that it was supposed to be compiled during the pip installation - which In my case would caused direct error during the installation. And not only during the first try run ...

**Comment by ProkopHapala** [comment]
@NikoOinonen I see there are conflicts in `Forces.h` and `ProbeParticle.cpp`. Is it due to my merge of damp func (https://github.com/Probe-Particle/ppafm/pull/87#pullrequestreview-1314533660)? If so, what was the proper way to do ?

**Comment by NikoOinonen** [comment]
> @NikoOinonen I see there are conflicts in `Forces.h` and `ProbeParticle.cpp`. Is it due to my merge of damp func ([#87 (review)](https://github.com/Probe-Particle/ppafm/pull/87#pullrequestreview-1314533660))? If so, what was the proper way to do ?

@ProkopHapala You did it right. We just happened to modify the same files in parallel a way that could not be automatically resolved. I just did a manual merge of the main branch into this branch, which resolved the conflict.

File: setup.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    package_path = os.path.dirname(os.path.realpath(__file__))
+    cpp_path = os.path.normpath(package_path + '/ppafm/cpp/')
+    current_directory = os.getcwd()
+    os.chdir (cpp_path)
+    os.system("make clean")
+    os.system("make all")
+    os.chdir (current_directory)
+
+def get_vars_path():
+    if not os.path.exists('C:\Program Files (x86)\Microsoft Visual Studio'):
```
Comment (yakutovicha) [inline setup.py:20]: Are there better ways to find out whether an executable exists?

File: ppafm/cpp_utils.py [hunk 2]
```diff
@@ -1,13 +1,33 @@
...
+    'PP': 'ProbeParticle',
+    'GU': 'GridUtils',
+    'fitting_': 'fitting',
+    'fitSpline_': 'fitSpline'
+}
 
 def work_dir( v__file__ ):
     return os.path.dirname( os.path.realpath( v__file__ ) )
 
 PACKAGE_PATH = work_dir( __file__ )
```
Comment (yakutovicha) [inline ppafm/cpp_utils.py:29]:
```suggestion
PACKAGE_PATH = Path(__file__).resolve().parent
```
Same for all the other occurrences.

File: ppafm/cpp_utils.py [hunk 3]
```diff
@@ -1,13 +1,33 @@
 import os
+import glob
+import platform
```
Comment (yakutovicha) [inline ppafm/cpp_utils.py:3]: Import pathlib:
```suggestion
import platform
from pathlib import Path 
```

File: setup.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    current_directory = os.getcwd()
+    os.chdir(cpp_path)
+    vars_path = get_vars_path()
+    for module in cpp_modules:
+        cmd = f'"{vars_path}" /no_logo /arch=amd64 && cl.exe /O2 /D_USRDLL /D_WINDLL {module}.cpp /link /dll /out:{module}_lib.dll'
+        print(cmd)
+        os.system(cmd)
+    os.chdir (current_directory)
+
+class Build(build):
```
Comment (yakutovicha) [inline setup.py:39]: I think it would look cleaner if we make the functions above members of the `Build` class.

**Review by yakutovicha — CHANGES_REQUESTED** [review]
Thanks, @NikoOinonen. A few comments and questions.

* What is the purpose of the `ppafm/dev` folder? You write that they are non-essential, why?
* I propose using [`Pathlib`](https://betterprogramming.pub/should-you-be-using-pathlib-6f3a0fddec7e) instead of `os.` (fine to do it in a follow-up PR, but please make an issue).

File: setup.py (hunk 1 above)
Comment (NikoOinonen) [reply setup.py:20]: Probably yes. I expect this to be quite fragile, but I just wanted to make something that works for now.

**Comment by NikoOinonen** [comment]
> What is the purpose of the ppafm/dev folder? You write that they are non-essential, why?

@yakutovicha We had some discussion with @ProkopHapala in https://github.com/Probe-Particle/ppafm/issues/75#issuecomment-1438472896 about the fact that these modules are something that are not essential to the core functionality of ppafm, and they are rather some additional functionality that Prokop intends to eventually move to his other project, so they are for now moved to this separate submodule.

File: pyproject.toml [hunk 5]
```diff
@@ -18,17 +18,20 @@ classifiers = [
 ]
 requires-python = ">=3.6"
 dependencies = [
-    "ase",
-    "matplotlib",
+    "matplotlib<=3.6.3",
```
[the remark this answers is not in the document]
Comment (NikoOinonen) [reply pyproject.toml:21]: Added a new issue #95.

**Comment by yakutovicha** [comment]
@ProkopHapala suggests deleting the `bin` folder from `ppafm/dev`.

File: pyproject.toml (hunk 5 above)
[the remark this answers is not in the document]
Comment (yakutovicha) [reply pyproject.toml:21]: Ok, please add something like the below to reference the issue:
```suggestion
    "matplotlib<=3.6.3",  # Remove upper limit when #95 is fixed.
```

File: setup.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+        if platform.system() == 'Windows':
+            self.build_windows()
+        else:
+            self.make()
+        super().run()
+
+    def make(self):
+        package_path = os.path.dirname(os.path.realpath(__file__))
+        cpp_path = os.path.normpath(package_path + '/ppafm/cpp/')
+        current_directory = os.getcwd()
```
Comment (yakutovicha) [inline setup.py:23]: I think it would be better to define those as a part of `__init__()` method, since you reuse them also in `build_windows`

**Review by yakutovicha — CHANGES_REQUESTED** [review]
A few more things I noticed after further testing.

File: ppafm/cpp_utils.py (hunk 2 above)
Comment (NikoOinonen) [reply ppafm/cpp_utils.py]: Change pathlib done.

File: setup.py (hunk 6 above)
Comment (NikoOinonen) [reply setup.py]: Done

File: ppafm/cpp/Vec3.h [hunk 7]
```diff
@@ -411,23 +411,23 @@ using Vec3d = Vec3T<double>;
...
-static constexpr Vec3f Vec3fX    = (Vec3f){1.0f,0.0f,0.0f};
-static constexpr Vec3f Vec3fY    = (Vec3f){0.0f,1.0f,0.0f};
-static constexpr Vec3f Vec3fZ    = (Vec3f){0.0f,0.0f,1.0f};
-
-static constexpr Vec3i Vec3iZero = (Vec3i){0,0,0};
-static constexpr Vec3i Vec3iOne  = (Vec3i){1,1,1};
-static constexpr Vec3i Vec3iX    = (Vec3i){1,0,0};
-static constexpr Vec3i Vec3iY    = (Vec3i){0,1,0};
-static constexpr Vec3i Vec3iZ    = (Vec3i){0,0,1};
+static constexpr Vec3d Vec3dZero = Vec3d {0.0,0.0,0.0};
```
[the remark this answers is not in the document]
Comment (NikoOinonen) [reply ppafm/cpp/Vec3.h:414]: Done.

File: ppafm/cpp/Vec2.h [hunk 8]
```diff
@@ -170,28 +170,28 @@ using Vec2i = Vec2T<int>;
 using Vec2f = Vec2T<float>;
 using Vec2d = Vec2T<double>;
```
[the remark this answers is not in the document]
Comment (NikoOinonen) [reply ppafm/cpp/Vec2.h:172]: Done.

**Comment by yakutovicha** [comment]
@NikoOinonen, for your information - I just merged the main branch into the current one.

**Review by yakutovicha — APPROVED** [review]
Looks good to me, thanks a lot @NikoOinonen.

Please merge it asap, I will then make the first working release on PyPi.

**Pull request merged**